test/test/xtdb/query_test.clj

```diff
@@ -1159,6 +1159,134 @@
                                    :where [[e :name "Ivan"]
                                            [(get-attr e :email nil) [email ...]]]})))))
 
+(t/deftest test-get-end-valid-time
+  (let [now #inst "2000-08-29T15:05:31.530-00:00"
+        later1 #inst "2050-08-29T15:05:31.530-00:00"
+        later2 #inst "2100-08-29T15:05:31.530-00:00"]
+    (fix/submit+await-tx [[::xt/put {:xt/id :foo, :type :some-value} now]
+                          [::xt/put {:xt/id :bar, :type :some-value2} now later2]])
+
+    (t/testing "existing end date at the edge"
+      (t/is (= #{[:foo nil]}
+               (xt/q (xt/db *api* now)
+                     '{:find [?e ?end-time]
+                       :where [[?e :type :some-value]
+                               [(get-end-valid-time ?e) ?end-time]]}))))
+
+    (t/testing "existing end date at the edge"
+      (t/is (= #{[:bar later2]}
+               (xt/q (xt/db *api* now)
+                     '{:find [?e ?end-time]
+                       :where [[?e :type :some-value2]
+                               [(get-end-valid-time ?e) ?end-time]]}))))
+
+    (t/testing "existing end date"
+      (t/is (= #{[:bar later2]}
+               (xt/q (xt/db *api* #inst "2025-08-29T15:05:31.530-00:00") ;; in-between
+                     '{:find [?e ?end-time]
+                       :where [[?e :type :some-value2]
+                               [(get-end-valid-time ?e) ?end-time]]}))))
+
+    (t/testing "existing and non-exisiting end dates"
+      (t/is (= #{[:foo nil] [:bar later2]}
+               (xt/q (xt/db *api* now)
+                     '{:find [?e ?end-time]
+                       :where [[?e :type]
+                               [(get-end-valid-time ?e) ?end-time]]}))))
+
+    (t/testing "non-exisiting entity"
+      (t/is (= #{}
+               (xt/q (xt/db *api* now)
+                     '{:find [?e ?end-time]
+                       :where [[?e :non-existing-type]
+                               [(get-end-valid-time ?e) ?end-time]]}))))
+
+    (t/testing "testing default value"
+      (t/is (= #{[:foo :the/default]}
+               (xt/q (xt/db *api* now)
+                     '{:find [?e ?end-time]
+                       :where [[?e :type :some-value]
+                               [(get-end-valid-time ?e :the/default) ?end-time]]}))))
+
+    (fix/submit+await-tx [[::xt/put {:xt/id :bar, :type :some-value2} later1 later2]])
+    (xt/sync *api*)
+
+    (t/testing "entity change"
+      (t/is (= #{[:bar later1]}
+               (xt/q (xt/db *api* now)
+                     '{:find [?e ?end-time]
+                       :where [[?e :type :some-value2]
+                               [(get-end-valid-time ?e) ?end-time]]})))
```

Comment:
If I understand your comment concerning `back in transaction` correctly. This test bit should already cover that. I added the `sync` call above it.

Reply:
paired on some tests to demo these cases